cryptography/src/bls12381/primitives/ops/aggregate.rs

```diff
@@ -0,0 +1,490 @@
+pub fn combine_public_keys<'a, V, I>(public_keys: I) -> PublicKey<V>
+where
+    V: Variant,
+    I: IntoIterator<Item = &'a V::Public>,
+    V::Public: 'a,
+{
+    let mut p = V::Public::zero();
+    for pk in public_keys {
+        p += pk;
+    }
+    PublicKey(p)
+}
+
+/// Combines multiple signatures into an aggregate signature.
+///
+/// # Warning
+///
+/// This function assumes a group check was already performed on each `signature` and
+/// that each `signature` is unique. If any of these assumptions are violated, an attacker can
+/// exploit this function to verify an incorrect aggregate signature.
+pub fn combine_signatures<'a, V, I>(signatures: I) -> Signature<V>
+where
+    V: Variant,
+    I: IntoIterator<Item = &'a V::Signature>,
+    V::Signature: 'a,
+{
+    let mut s = V::Signature::zero();
+    for sig in signatures {
+        s += sig;
+    }
+    Signature(s)
+}
+
+/// Verifies the aggregate signature over a single message from multiple public keys.
+///
+/// # Warning
+///
+/// This function assumes the caller has performed a group check and collected a proof-of-possession
+/// for all provided `public`. This function assumes a group check was already performed on the
+/// `signature`. It is not safe to provide duplicate public keys.
+pub fn verify_multiple_public_keys<'a, V, I>(
+    public: I,
+    namespace: Option<&[u8]>,
+    message: &[u8],
+    signature: &Signature<V>,
+) -> Result<(), Error>
+where
+    V: Variant,
+    I: IntoIterator<Item = &'a V::Public>,
+    V::Public: 'a,
+{
+    // Aggregate public keys
+    //
+    // We can take advantage of the bilinearity property of pairings to aggregate public keys
+    // that have all signed the same message (as long as all public keys are unique).
+    let agg_public = combine_public_keys::<V, _>(public);
+
+    // Compute the hash of the message
+    let hm = namespace.map_or_else(
+        || hash_message::<V>(V::MESSAGE, message),
+        |ns| hash_message_namespace::<V>(V::MESSAGE, ns, message),
+    );
+
+    // Verify the signature
+    V::verify(agg_public.inner(), &hm, signature.inner())
+}
+
+/// Verifies an aggregate signature over multiple unique messages from a single public key.
+///
+/// Each entry is a tuple of (namespace, message). The signature must be the aggregate
+/// of all individual signatures.
+///
+/// # Warning
+///
+/// This function is vulnerable to signature malleability when used with signatures
+/// that were aggregated from different messages. An attacker can redistribute
+/// signature components between messages while keeping the aggregate unchanged.
+/// Use [`batch::verify_multiple_messages`](super::batch::verify_multiple_messages) instead when signatures are provided individually.
+///
+/// This function assumes a group check was already performed on `public` and `signature`.
+/// It is not safe to provide an aggregate public key or to provide duplicate messages.
+pub fn verify_multiple_messages<'a, V, I>(
```

Comment:
We should remove this method completely. Signature aggregation should only be used in the case where you want to aggregate signatures from separate entities, not as "batched verification with tricky caveats"

Reply:
This function was added to support #1353 (compressed nullifications). While we have been delayed in implementing it, I do think it is indicative of there being usefulness to this method and wish to keep it for now.

I'll make sure to update the documentation.